File: Pinocchio/Pinocchio/Model/Actor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework;
using Pinocchio.animation3D;

namespace Pinocchio.Model
{
    class Actor
    {
        ArrayList boneList;
        Dictionary<Bone.BoneType, Bone> boneMap;
        Vector3 position = new Vector3();
        Vector3 rotation = new Vector3();
        Vector3 scale = new Vector3(1.0f);
        Animation3DInstance curAnimation = new Animation3DInstance(null);

        public Actor(GraphicsDevice graphicsDevice)
        {
            // 본리스트 초기화
            initializeBones(graphicsDevice);
            
        }

        private void initializeBones(GraphicsDevice graphicsDevice)
        {
            // 초기화
            const float baseSize = 10.0f;
            boneList = new ArrayList();
            boneMap = new Dictionary<Bone.BoneType,Bone>();

            /// 본 추가
            addBone(graphicsDevice, Bone.BoneType.None, new Vector3(0f,0f,0f), Quaternion.Identity, new Vector3(1.0f),
                "Hip_Center", Bone.BoneType.Hip_Center, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Hip_Center, new Vector3(0f, 10f, 0f), Quaternion.Identity, new Vector3(1.0f),
                "Spine", Bone.BoneType.Spine, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Spine, new Vector3(0f, 40f, 0f), Quaternion.Identity, new Vector3(1.0f),
                "Shoulder_Center", Bone.BoneType.Shoulder_Center, new Vector3(baseSize), Vector3.Zero);

            addBone(graphicsDevice, Bone.BoneType.Shoulder_Center, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Head", Bone.BoneType.Head, new Vector3(baseSize), Vector3.Zero);

            addBone(graphicsDevice, Bone.BoneType.Shoulder_Center, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Shoulder_Right", Bone.BoneType.Shoulder_Right, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Shoulder_Right, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Elbow_Right", Bone.BoneType.Elbow_Right, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Elbow_Right, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Wrist_Right", Bone.BoneType.Wrist_Right, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Wrist_Right, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Hand_Right", Bone.BoneType.Hand_Right, new Vector3(baseSize), Vector3.Zero);

            addBone(graphicsDevice, Bone.BoneType.Shoulder_Center, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Shoulder_Left", Bone.BoneType.Shoulder_Left, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Shoulder_Left, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Elbow_Left", Bone.BoneType.Elbow_Left, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Elbow_Left, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Wrist_Left", Bone.BoneType.Wrist_Left, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Wrist_Left, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Hand_Left", Bone.BoneType.Hand_Left, new Vector3(baseSize), Vector3.Zero);

            addBone(graphicsDevice, Bone.BoneType.Hip_Center, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Hip_Right", Bone.BoneType.Hip_Right, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Hip_Right, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Knee_Right", Bone.BoneType.Knee_Right, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Knee_Right, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Ankle_Right", Bone.BoneType.Ankle_Right, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Ankle_Right, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Foot_Right", Bone.BoneType.Foot_Right, new Vector3(baseSize), Vector3.Zero);

            addBone(graphicsDevice, Bone.BoneType.Hip_Center, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Hip_Left", Bone.BoneType.Hip_Left, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Hip_Left, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Knee_Left", Bone.BoneType.Knee_Left, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Knee_Left, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Ankle_Left", Bone.BoneType.Ankle_Left, new Vector3(baseSize), Vector3.Zero);
            addBone(graphicsDevice, Bone.BoneType.Ankle_Left, new Vector3(), Quaternion.Identity, new Vector3(1.0f),
                "Foot_Left", Bone.BoneType.Foot_Left, new Vector3(baseSize), Vector3.Zero);

        }

        private void addBone(GraphicsDevice graphicsDevice, Bone.BoneType parentType, Vector3 localPosition, Quaternion localRotation, Vector3 localScale,
            string name, Bone.BoneType type, Vector3 size, Vector3 center)
        {
            // 본을 생성
            Bone parentBone = null;
            if (boneMap.ContainsKey(parentType))
                parentBone = boneMap[parentType];
            Bone bone = new Bone(graphicsDevice, parentBone, name, type, size, center);

            // 본의 부모상대 변환행렬 설정
            Matrix localTransform = Matrix.CreateScale(localScale) * Matrix.CreateFromQuaternion(localRotation) * Matrix.CreateTranslation(localPosition);
            bone.LocalTransform = localTransform;

            // 리스트와 맵에 추가
            boneList.Add(bone);
            boneMap[bone.Type] = bone;
        }

        public void update(int deltaFrame)
        {
            // 애니메이션 업데이트
            curAnimation.update(boneList, deltaFrame);
        }

        public void draw(Matrix view, Matrix projection, Color color)
        {
            // 각 본들을 그린다.
            //foreach (Bone bone in boneList)
            //    bone.draw(view, projection, color);
            for (int i = 0; i < 3; ++i)
            {
                ((Bone)boneList[i]).draw(view, projection, color);
            }
        }

        public void setCurAnimation(Animation3D animation)
        {
            curAnimation = new Animation3DInstance(animation);
        }


        // 접근
        public Vector3 Position { get { return position; } set { position = value; } }
        public Vector3 Rotation { get { return rotation; } set { rotation = value; } }
        public Vector3 Scale { get { return scale; } set { scale = value; } }
    }

}

File: MotionCapturer/MotionCapturer/source/model/Actor.cpp

#include "stdafx.h"
#include "Actor.h"

#include <GL/glew.h>

#include "../animation/AnimationInstance.h"
#include "Bone.h"


Actor::Actor() : animationInstance(null)
{
	animationInstance = new AnimationInstance;

	scale.set(0.01f, 0.01f, 0.01f);

    // 본리스트 초기화
    initializeBones();
}

Actor::~Actor()
{
	destroy();
}

void Actor::destroy()
{
	for (uint i=0; i<boneList.size(); ++i)
	{
		DELETE(boneList[i])
	}
	boneList.clear();
	boneMap.clear();
}

void Actor::addBone(int parentType, const Vector3 &localPosition, const Quaternion &localRotation, const Vector3 &localScale,
        string name, int type)
{
    // 본을 생성
    Bone *parentBone = null;
	if (boneMap.find(parentType) != boneMap.end())
        parentBone = boneMap[parentType];
	Bone *bone = new Bone(parentBone, name, (Bone::BoneType)type);

	// 상대 위치 설정
	bone->setLocalPosition(localPosition);

    // 본의 부모상대 변환행렬 설정
	Matrix transMat, rotMat, scaleMat, localTransform;
	transMat.setTranslate(localPosition);
	rotMat.setRotate(localRotation);
	scaleMat.setScale(localScale);
	localTransform.multiply(transMat, rotMat);
	localTransform.multiply(scaleMat);
	bone->setTransformFromParent(localTransform);

    // 리스트와 맵에 추가
	boneList.push_back(bone);
	boneMap[bone->getBoneType()] = bone;
}

void Actor::initializeBones()
{
    // 초기화
    const float baseSize = 10.f;
    destroy();

    /// 본 추가
    addBone(Bone::None, Vector3(0.f,0.f,0.f), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Hip_Center", Bone::Hip_Center);
    addBone(Bone::Hip_Center, Vector3(0.f, 20.f, 0.f), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Spine", Bone::Spine);
    addBone(Bone::Spine, Vector3(0.f, 50.f, 0.f), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Shoulder_Center", Bone::Shoulder_Center);

    addBone(Bone::Shoulder_Center, Vector3(0, 25, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Head", Bone::Head);

    addBone(Bone::Shoulder_Center, Vector3(-30, -10, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Shoulder_Right", Bone::Shoulder_Right);
    addBone(Bone::Shoulder_Right, Vector3(-50, 0, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Elbow_Right", Bone::Elbow_Right);
    addBone(Bone::Elbow_Right, Vector3(-40, 0, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Wrist_Right", Bone::Wrist_Right);
    addBone(Bone::Wrist_Right, Vector3(-20, 0, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Hand_Right", Bone::Hand_Right);

    addBone(Bone::Shoulder_Center, Vector3(30, -10, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Shoulder_Left", Bone::Shoulder_Left);
    addBone(Bone::Shoulder_Left, Vector3(50, 0, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Elbow_Left", Bone::Elbow_Left);
    addBone(Bone::Elbow_Left, Vector3(40 , 0, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Wrist_Left", Bone::Wrist_Left);
    addBone(Bone::Wrist_Left, Vector3(20, 0, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Hand_Left", Bone::Hand_Left);

    addBone(Bone::Hip_Center, Vector3(-20, -10, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Hip_Right", Bone::Hip_Right);
    addBone(Bone::Hip_Right, Vector3(0, -90, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Knee_Right", Bone::Knee_Right);
    addBone(Bone::Knee_Right, Vector3(0, -70, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Ankle_Right", Bone::Ankle_Right);
    addBone(Bone::Ankle_Right, Vector3(0, -10, -20), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Foot_Right", Bone::Foot_Right);

    addBone(Bone::Hip_Center, Vector3(20, -10, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Hip_Left", Bone::Hip_Left);
    addBone(Bone::Hip_Left, Vector3(0, -90, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Knee_Left", Bone::Knee_Left);
    addBone(Bone::Knee_Left, Vector3(0, -70, 0), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Ankle_Left", Bone::Ankle_Left);
    addBone(Bone::Ankle_Left, Vector3(0, -10, -20), Quaternion(), Vector3(1.f, 1.f, 1.f),
        "Foot_Left", Bone::Foot_Left);

}

    
void Actor::update(int deltaFrame)
{
    // 애니메이션 업데이트
    animationInstance->update(boneList, deltaFrame);
}

void Actor::draw()
{
	glPushMatrix();

	// 변환행렬 설정
	glTranslatef(position.getX(), position.getY(), position.getZ());
	glRotatef(rotation.getY(), 0.f, 1.f, 0.f);
	glRotatef(rotation.getX(), 1.f, 0.f, 0.f);
	glRotatef(rotation.getZ(), 0.f, 0.f, 1.f);
	glScalef(scale.getX(), scale.getY(), scale.getZ());
	
	glPointSize(10.f);
	glColor4ub(255, 255, 0, 255);
	
	for (uint i=0; i<boneList.size(); ++i)
	{
		glPushMatrix();

		glMultMatrixf(boneList[i]->getTransform().getPointer());
			
		glBegin(GL_POINTS);
			glVertex3f(0,0,0);
		glEnd();
			
		glPopMatrix();
	}

	glPopMatrix();
}

void Actor::setCurAnimation(Animation *animation)
{
	animationInstance->setAnimation(animation);
}
